server/subscribe.go

```diff
@@ -22,10 +22,16 @@ const (
 	JIRA_SUBSCRIPTIONS_KEY = "jirasub"
 )
 
+type SubscriptionFilters struct {
```

Comment:
Could we actually store these as sets too? We could do write some custom json serialization/de-serialization if we really want to store them as plain slices.

Reply:
I'm not sure I understand. You want to continue storing them in the KVStore as json arrays (which makes sense because that is what we serve to the UI), but deserialize them into Sets upon parsing? It sounds useful, but I'm not familiar with how you handle this.

Is this what you have in mind @crspeller?

Reply:
Yup exactly. You can create a custom marshal/unmarshal. Example here: https://golang.org/pkg/encoding/json/#example__customMarshalJSON

Reply:
Wow that's so cool! Sounds have RTM, thanks!

Reply:
Sets are useful. I agree, I do _not_ fully understand the logic of no having at least some basic set types in go (I do not step into the generics debates). BTW, for all our intents using sorted slices for sets is probably fine. Remember, the only operations we really need are Union and Intersect, and these are O(n) with sorted slices, no worse than maps. And for making them, they are small enough to sort once?

Reply:
@levb
Something I have noticed about the design here is that we would be avoiding mutating an existing set, and instead want to create a new one (specifically with `Add`). Would the set not serve better as a mutable data structure? It doesn't necessarily have to be behind the scenes, but what about making copies is beneficial versus mutating? I am thinking of maps, because it automatically keeps things unique through adding/overwriting keys, and provides simple, quick lookup.

Also, do you mind explaining why using the Set as a parameter and not as a receiver is the better option?

Reply:
immutable vs mutable - immutable among other things, allows for compiler to make sense of things, and avoids issues with concurrency. It's simpler. BTW, this API does not imply immutability, much like append may mutate the underlying array of a slice, but still return a new slice reference.

Function vs Method: Esthetics? The operations are symmetrical, using a method implies a certain asymmetry. But I am 0/5 on that, making it a method on a value receiver is literally no different. You can even use an a.Method as a func (a A).

Reply:
^^ Got me thinking that `s/Add(a, elems...)/Append(a, elems...)/` to follow `append()`. That's as _idiomatic_ as it gets.